### ISSUE TITLE
Should the `metricdata.HistogramDataPoint` be generic?

### ISSUE THREAD

**Issue by MrAlias** [open-telemetry/opentelemetry-go#3834]
Currently, the `metricdata.HistogramDataPoint` is defined without any generic argument.

https://github.com/open-telemetry/opentelemetry-go/blob/c869b9171ac19f2321bd685fe32199c3ad668418/sdk/metric/metricdata/data.go#L106-L129

However, when we add exemplars (https://github.com/open-telemetry/opentelemetry-go/issues/559), this `struct` will contain exemplars that likely have a typed measurement value. For example:

```go
type HistogramDataPoint struct {
	// ...
	Exemplars []Exemplar[N] // Invalid
}

type Exemplar[N int64 | float64] struct {
	Timestamp time.Time
	SpanContext trace.SpanContext
	Value N
	Attributes []attribute.KeyValue
}
```

We could also define an abstraction over the `Exemplar[N]`, like this:

```go
type HistogramDataPoint struct {
	// ...
	Exemplars Exemplars
}

type Exemplars []ExemplarValue

type ExemplarValue interface {
	privateExemplarValue()
}

type Exemplar[N int64 | float64] struct {
	Timestamp time.Time
	SpanContext trace.SpanContext
	Value N
	Attributes []attribute.KeyValue
}

func (Exemplar[N]) privateExemplarValue() {}
```

- This would mean there is an added type needed to "privatize" and abstract both `Exemplar[N]`
- There will need to be type switching when decoding this to OTLP
- It would allow both `Exemplar[int64]` and `Exemplar[float64]` to be included in the same `Exemplars`
- This would still allow `metricdata.DataPoint[N]` to have a `Exemplars []Exemplar[N]` field added
  - This would then mean that the `Exemplar` for `metricdata.HistogramDataPoint` would not be defined the same way as `metricdata.DataPoint[N]`

Alternatively, the `Value` of the `Exemplar` could be abstracted with an interface:

```go
type Exemplar struct {
	Timestamp time.Time
	SpanContext trace.SpanContext
	Value ExemplarValue
	Attributes []attribute.KeyValue
}

type ExemplarValue interface {
	privateExemplarValue()
}

type Measurement[N int64 | float64] struct {
	Value N
}

func (Measurement[N]) privateExemplarValue() {}
```

- This means that the `metricdata.Datapoint` would not be able to pass through its generic argument
  - A second type assertion will be need to decode the `Exemplar.Value`

**Comment by pellared** [comment]
I think that we could have 

```go
type Measurement interface {
	int64 | float64
}
```

we cold be even more abstract (but I do not think it is a good idea - I think it would over-abstraction)

```go
type Measurement interface {
	~int64 | ~float64
}
```

Inspired by: https://pkg.go.dev/golang.org/x/exp/constraints

**Comment by pellared** [comment]
Yes, it is only about defining a constrain for the measurement value type. E.g.

```go
type Measurement interface {
	int64 | float64
}

type Exemplar[N Measurement] struct {
	Timestamp time.Time
	SpanContext trace.SpanContext
	Value N
	Attributes []attribute.KeyValue
}
```

The reason behind is that it would be easier to add more supported value types in future e.g. `int32`.

However this approach will still not help if we would like to support types from https://pkg.go.dev/math/big package.

**Comment by MrAlias** [comment]
Closed by #3849
